File: dart_lang/lib/a_solid/d_interface_segregation/a_pre_refactor/customer.dart
import 'package:dart_lang/a_solid/d_interface_segregation/a_pre_refactor/user_manager.dart';

class Customer implements UserManager {
  String? _name;
  String? _email;

  Customer(String name) : _name = name;

  String? get email => _email;
  String? get name => _name;

  @override
  void changePassword(Customer customer, String newPassword) {
    print("Changing password ...");
  }

  @override
  void subscribeToMSNotifications() {
    print("Notifying password ...");
  }

  @override
  void subscribeToNewProductsAvailability() {
    print("subscribing ...");
  }

  @override
  void updateUserProfile(Customer customer) {
    print("updating");
  }
}

File: PHP/Solid Principles/interface_segregation/problem/Customer.php
<?php

class Customer implements UserManager {
    private ?string $_name;
    private ?string $_email;

    public function __construct(string $name) {
        $this->_name = $name;
    }

    public function getEmail(): ?string {
        return $this->_email;
    }

    public function getName(): ?string {
        return $this->_name;
    }

    public function changePassword(Customer $customer, string $newPassword): void {
        echo "Changing password for customer: " . $customer->getName();
    }

    public function subscribeToMSNotifications(Customer $customer): void {
        echo "Subscribing customer to MS Notifications";
    }

    public function subscribeToNewProductsAvailability(Customer $customer): void {
        echo "Subscribing customer to new product notifications";
    }

    public function updateUserProfile(Customer $customer): void {
        echo "Updating customer profile";
    }
}